drivers/auxdisplay/auxdisplay_gpio_7seg.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2024 Chen Xingyu <hi@xingrz.me>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_7_segment
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/auxdisplay.h>
+#include <zephyr/drivers/gpio.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(auxdisplay_gpio_7seg, CONFIG_AUXDISPLAY_LOG_LEVEL);
+
+static const uint8_t DIGITS[] = {
+	[0] = BIT(0) | BIT(1) | BIT(2) | BIT(3) | BIT(4) | BIT(5),
+	[1] = BIT(1) | BIT(2),
+	[2] = BIT(0) | BIT(1) | BIT(3) | BIT(4) | BIT(6),
+	[3] = BIT(0) | BIT(1) | BIT(2) | BIT(3) | BIT(6),
+	[4] = BIT(1) | BIT(2) | BIT(5) | BIT(6),
+	[5] = BIT(0) | BIT(2) | BIT(3) | BIT(5) | BIT(6),
+	[6] = BIT(0) | BIT(2) | BIT(3) | BIT(4) | BIT(5) | BIT(6),
+	[7] = BIT(0) | BIT(1) | BIT(2),
+	[8] = BIT(0) | BIT(1) | BIT(2) | BIT(3) | BIT(4) | BIT(5) | BIT(6),
+	[9] = BIT(0) | BIT(1) | BIT(2) | BIT(3) | BIT(5) | BIT(6),
+};
+
+#define DP    (BIT(7))
+#define BLANK (0x00)
+
+struct auxdisplay_gpio_7seg_data {
+	struct k_timer timer;
+	uint32_t current_digit;
+};
+
+struct auxdisplay_gpio_7seg_config {
+	struct auxdisplay_capabilities capabilities;
+	const struct gpio_dt_spec *segment_gpios;
+	const struct gpio_dt_spec *digit_gpios;
+	uint32_t segment_count;
+	uint32_t digit_count;
+	uint32_t refresh_period_ms;
+	uint8_t *buffer;
+};
+
+static int auxdisplay_gpio_7seg_display_on(const struct device *dev)
+{
+	const struct auxdisplay_gpio_7seg_config *cfg = dev->config;
+	struct auxdisplay_gpio_7seg_data *data = dev->data;
+
+	data->current_digit = 0;
+	k_timer_start(&data->timer, K_NO_WAIT, K_MSEC(cfg->refresh_period_ms));
+
+	return 0;
+}
+
+static int auxdisplay_gpio_7seg_display_off(const struct device *dev)
+{
+	struct auxdisplay_gpio_7seg_data *data = dev->data;
+
+	k_timer_stop(&data->timer);
+
+	return 0;
+}
+
+static int auxdisplay_gpio_7seg_capabilities_get(const struct device *dev,
+						 struct auxdisplay_capabilities *capabilities)
+{
+	const struct auxdisplay_gpio_7seg_config *cfg = dev->config;
+
+	memcpy(capabilities, &cfg->capabilities, sizeof(struct auxdisplay_capabilities));
+
+	return 0;
+}
+
+static int auxdisplay_gpio_7seg_clear(const struct device *dev)
+{
+	const struct auxdisplay_gpio_7seg_config *cfg = dev->config;
+	struct auxdisplay_gpio_7seg_data *data = dev->data;
+
+	memset(cfg->buffer, 0, cfg->digit_count);
+	data->current_digit = 0;
+
+	return 0;
+}
+
+static int auxdisplay_gpio_7seg_write(const struct device *dev, const uint8_t *ch, uint16_t len)
+{
+	const struct auxdisplay_gpio_7seg_config *cfg = dev->config;
+	struct auxdisplay_gpio_7seg_data *data = dev->data;
+	uint32_t ib, ic;
+
+	for (ib = 0, ic = 0; ib < cfg->digit_count && ic < len; ic++) {
+		if (ch[ic] >= '0' && ch[ic] <= '9') {
+			cfg->buffer[ib++] = DIGITS[ch[ic] - '0'];
+		} else if (ch[ic] == '.') {
```

Comment:
This does not check that a dot segment is present

Reply:
It's okay here since it's just setting the buffer bits and not actually updating the GPIOs directly. The `DP` bit will be ignored if only 7 GPIOs are present, as handled in the `auxdisplay_gpio_7seg_timer_expiry_fn` function below.